glotaran/builtin/megacomplexes/pfid/pfid_megacomplex.py

```diff
@@ -0,0 +1,273 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import numpy as np
+import xarray as xr
+from scipy.special import erf
+
+from glotaran.builtin.megacomplexes.decay.irf import Irf
+from glotaran.builtin.megacomplexes.decay.irf import IrfMultiGaussian
+from glotaran.model import DatasetModel
+from glotaran.model import ItemIssue
+from glotaran.model import Megacomplex
+from glotaran.model import Model
+from glotaran.model import ModelItemType
+from glotaran.model import ParameterType
+from glotaran.model import attribute
+from glotaran.model import item
+from glotaran.model import megacomplex
+
+if TYPE_CHECKING:
+    from glotaran.parameter import Parameters
+    from glotaran.typing.types import ArrayLike
+
+
+class OscillationParameterIssue(ItemIssue):
+    def __init__(self, label: str, len_labels: int, len_frequencies: int, len_rates: int):
+        self._label = label
+        self._len_labels = len_labels
+        self._len_frequencies = len_frequencies
+        self._len_rates = len_rates
+
+    def to_string(self) -> str:
+        return (
+            f"Size of labels ({self.len_labels}), frequencies ({self.len_frequencies}) "
+            f"and rates ({self.len_rates}) does not match for damped oscillation "
+            f"megacomplex '{self.label}'."
+        )
+
+
+def validate_pfid_parameter(
+    labels: list[str],
+    pfid: PFIDMegacomplex,
+    model: Model,
+    parameters: Parameters | None,
+) -> list[ItemIssue]:
+    issues = []
+
+    len_labels, len_frequencies, len_rates = (
+        len(pfid.labels),
+        len(pfid.frequencies),
+        len(pfid.rates),
+    )
+
+    if len({len_labels, len_frequencies, len_rates}) > 1:
+        issues.append(
+            OscillationParameterIssue(pfid.label, len_labels, len_frequencies, len_rates)
+        )
+
+    return issues
+
+
+@item
+class PFIDDatasetModel(DatasetModel):
+    spectral_axis_inverted: bool = False
+    spectral_axis_scale: float = 1
+    irf: ModelItemType[Irf] | None = None
+
+
+@megacomplex(dataset_model_type=PFIDDatasetModel)
+class PFIDMegacomplex(Megacomplex):
+    dimension: str = "time"
+    type: str = "pfid"
+    labels: list[str] = attribute(validator=validate_pfid_parameter)
+    frequencies: list[ParameterType]  # omega_a
+    rates: list[ParameterType]  # 1/T2
+
+    def calculate_matrix(
+        self,
+        dataset_model: DatasetModel,
+        global_axis: ArrayLike,
+        model_axis: ArrayLike,
+        **kwargs,
+    ):
+        clp_label = [f"{label}_cos" for label in self.labels] + [
+            f"{label}_sin" for label in self.labels
+        ]
+
+        frequencies = np.array(self.frequencies)
+        rates = np.array(self.rates)
+
+        if dataset_model.spectral_axis_inverted:
+            frequencies = dataset_model.spectral_axis_scale / frequencies
+        elif dataset_model.spectral_axis_scale != 1:
```

Comment:
Shouldn't this be an `if` to also work with `dataset_model.spectral_axis_inverted==True`?

Reply:
It's the same in the spectral megacomplex, but we may reinvestigate in the future.